Adding comments on Kanister Resource Types Fields in types.go

Kartik-Garg opened the pull request:
There were comments missing from kanister resource types fields in types.go ,because of which there was missing description while creating the documentation using tools like https://github.com/ahmetb/gen-crd-api-reference-docs/ . So, added missing comments on fields present in types.go

Signed-off-by: Kartik-Garg <kartikgarg938@gmail.com>

## Change Overview
The documentation generated now has description of all the fields present in Types.go:
![Screenshot from 2022-11-21 20-49-36](https://user-images.githubusercontent.com/25133864/203091866-9c8079a4-a580-42bd-8893-2f39f21f8430.png)

<!-- Insert PR description-->
Added missing comments on fields on types.go file.
## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [x] :world_map: Documentation
- [ ] :robot: Test

## Issues <!-- to auto-close the issue, add the "fixes" keyword -->

- fixes #1721 

## Test Plan

<!-- Will run prior to merging.-->
<!-- Include example how to run.-->

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
+	SecretNames []string `json:"secretNames,omitempty"`
+	// InputArtifactNames is the array of artifacts used in BlueprintAction.
+	InputArtifactNames []string `json:"inputArtifactNames,omitempty"`
+	// OutputArtifacts is the map of artifacts which is received from BlueprintAction.
```
```suggestion
	//  The map of rendered parameters made available to the BlueprintAction.
```

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
+	ConfigMapNames []string `json:"configMapNames,omitempty"`
+	// SecretNames is the array of secrets used in BlueprintAction.
+	SecretNames []string `json:"secretNames,omitempty"`
+	// InputArtifactNames is the array of artifacts used in BlueprintAction.
```
List of Artifact names used in the action phases.

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
+	InputArtifactNames []string `json:"inputArtifactNames,omitempty"`
+	// OutputArtifacts is the map of artifacts which is received from BlueprintAction.
+	OutputArtifacts map[string]Artifact `json:"outputArtifacts,omitempty"`
+	// Phases is the array of phases that constitute BlueprintAction.
```
```suggestion
	//  List of BlueprintPhases which are invoked in order when executing this Action.
```

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
+	OutputArtifacts map[string]Artifact `json:"outputArtifacts,omitempty"`
+	// Phases is the array of phases that constitute BlueprintAction.
+	Phases []BlueprintPhase `json:"phases,omitempty"`
+	// DeferPhase is the phase which is present in the Blueprint.
```
```suggestion
	// DeferPhase invoked after the execution of Phases defined above. A DeferPhase, when specified, is executed regardless of the statuses of the Phases. A DeferPhase can be used for cleanup operations at the end of an Action.
```

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
 type BlueprintPhase struct {
-	Func       string                     `json:"func"`
-	Name       string                     `json:"name"`
+	// Func is the function present in the BluePrintPhase.
```
```suggestion
	// Func is the name of a registered Kanister function.
```

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
-	Name       string                     `json:"name"`
+	// Func is the function present in the BluePrintPhase.
+	Func string `json:"func"`
+	// Name is the name of BlueprintPhase.
```
```suggestion
	// Name of the phase.
```

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
+	Func string `json:"func"`
+	// Name is the name of BlueprintPhase.
+	Name string `json:"name"`
+	// ObjectRefs is the object referent.
```
```suggestion
	// A map of references to the Kubernetes objects on which the action will be performed.
```

PrasadG193 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,40 @@ var _ runtime.Object = (*Blueprint)(nil)
+	// ObjectRefs is the object referent.
 	ObjectRefs map[string]ObjectReference `json:"objects,omitempty"`
-	Args       map[string]interface{}     `json:"args"`
+	// Args are the arguments used in the BlueprintPhase
```
```suggestion
	// A map of named arguments that the controller will pass to the Kanister function.
```

PrasadG193 reviewed (commented):
@Kartik-Garg Please go through https://docs.kanister.io/architecture.html to understand the significance of each field and correct the comments. I went through few but will re-review once you correct the field comments by referring the doc

Kartik-Garg commented:
> @Kartik-Garg Please go through https://docs.kanister.io/architecture.html to understand the significance of each field and correct the comments. I went through few but will re-review once you correct the field comments by referring the doc

@PrasadG193 I see, I did not realize we could write comments by referring to https://docs.kanister.io/architecture.html 
Thanks a lot for pointing it out :D

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -51,8 +51,10 @@ var _ runtime.Object = (*ActionSet)(nil)
 	metav1.ObjectMeta `json:"metadata"`
-	Spec              *ActionSetSpec   `json:"spec,omitempty"`
-	Status            *ActionSetStatus `json:"status,omitempty"`
+	// Spec is the spec field present in the ActionSet.
```
Modified the comment with reference to https://docs.kanister.io/architecture.html

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
-	LocationTypeGCS         LocationType = "gcs"
+	// LocationType is selected as GCS.
+	LocationTypeGCS LocationType = "gcs"
+	// LocationType is selected as S3 Complaint.
 	LocationTypeS3Compliant LocationType = "s3Compliant"
-	LocationTypeAzure       LocationType = "azure"
-	LocationTypeKopia       LocationType = "kopia"
+	// LocationType is selected as Azure.
+	LocationTypeAzure LocationType = "azure"
+	// LocationType is selected as Kopia.
+	LocationTypeKopia LocationType = "kopia"
```
do we have to add comments here as well. If it's not mandatory let's not add comment to  these. Because of the same reason that it's very self explanatory.

Kartik-Garg replied:
Alright, removing comments from here.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
-	Region   string       `json:"region"`
+	// Type of the Location being used.
+	Type LocationType `json:"type"`
+	// Represents the bucket being used for the current Location.
```
Bucket being used, or bucket that should be used, or going to be used?

viveksinghggits replied:
Bucket represents the bucket on the object storage where the backup would be uploaded to.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
+	// Type of the Location being used.
+	Type LocationType `json:"type"`
+	// Represents the bucket being used for the current Location.
+	Bucket string `json:"bucket"`
+	// Represents the endpoints being used by the current Location.
+	Endpoint string `json:"endpoint"`
+	// Prefix is the string used in the beginning of the Location.
+	Prefix string `json:"prefix"`
+	// Represents the region of the current location.
+	Region string `json:"region"`
```
This comments are incorrect I think. Let us know the fields that you are unsure about and we can give the details about those fields.

viveksinghggits replied:
Region represents the region where the bucket, mentioned above, is present in

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -109,9 +112,13 @@ type ActionSpec struct {
-	State    State          `json:"state"`
-	Actions  []ActionStatus `json:"actions,omitempty"`
-	Error    Error          `json:"error,omitempty"`
+	// State represents the current state of the ActionSet.
```
Alright, got it.
Will be making changes in the comments in the suggested format.

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,42 @@ var _ runtime.Object = (*Blueprint)(nil)
-	DeferPhase         *BlueprintPhase     `json:"deferPhase,omitempty"`
+	// Name of the action.
+	Name string `json:"name"`
+	// Represents the type of Kubernetes object this BlueprintAction is written for.
+	Kind string `json:"kind"`
```
I have added `Kind field is now deprecated.`

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,42 @@ var _ runtime.Object = (*Blueprint)(nil)
+	Name string `json:"name"`
+	// Represents the type of Kubernetes object this BlueprintAction is written for.
+	Kind string `json:"kind"`
+	// List of Kubernetes ConfigMap names referred to in the action phases.
+	ConfigMapNames []string `json:"configMapNames,omitempty"`
```
changed it with `ConfigMapNames is used to specify the config map names that can be used later in the Action phases.`

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,42 @@ var _ runtime.Object = (*Blueprint)(nil)
+	ConfigMapNames []string `json:"configMapNames,omitempty"`
+	// List of Kubernetes secret names used in action phases.
+	SecretNames []string `json:"secretNames,omitempty"`
+	// InputArtifactNames is the list of Artifact names used in action phases.
+	InputArtifactNames []string `json:"inputArtifactNames,omitempty"`
```
Changed it accordingly

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +210,42 @@ var _ runtime.Object = (*Blueprint)(nil)
+	SecretNames []string `json:"secretNames,omitempty"`
+	// InputArtifactNames is the list of Artifact names used in action phases.
+	InputArtifactNames []string `json:"inputArtifactNames,omitempty"`
+	// OutputArtifacts is the map of rendered parameters made available to the BlueprintAction.
+	OutputArtifacts map[string]Artifact `json:"outputArtifacts,omitempty"`
```
yup, makes sense. 
changed it according to the suggestion

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
+	Location Location `json:"location"`
+	// Represents the credentials associated with the Location.
+	Credential Credential `json:"credential"`
+	// SkipSSLVerify is boolean and specifies whether skipping SkipSSLVerify verification
+	// is allowed when operating with the Location.
+	// If omitted from a CR definition it default to false
+	SkipSSLVerify bool `json:"skipSSLVerify"`
```
I found it written here https://docs.kanister.io/architecture.html, it is mentioned under profiles as: 
`SkipSSLVerify is boolean and specifies whether skipping SkipSSLVerify verification is allowed when operating with the Location. If omitted from a CR definition it default to false`

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -109,9 +112,13 @@ type ActionSpec struct {
+	Actions []ActionStatus `json:"actions,omitempty"`
+	// Error details for the failed ActionSet.
+	Error Error `json:"error,omitempty"`
+	// Progress provides information about the progress of an action.
```
Currently referring to https://docs.kanister.io/architecture.html for writing comments on the fields.
Couldn't find `progress` field in the document

viveksinghggits replied:
```
Progress provides information of how much progress has been made for an actionset. for example the percentage an actionset has been completed and the phase that is being run currently for an actionset.
```

try to make the grammar better before pushing this :).

Kartik-Garg replied:
Thanks for the response, made the changes accordingly.

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
-	Location          Location   `json:"location"`
-	Credential        Credential `json:"credential"`
-	SkipSSLVerify     bool       `json:"skipSSLVerify"`
+	// Specifies the location that the Blueprint can use.
```
Alright, got it
I had referred to https://docs.kanister.io/architecture.html#profiles while writing this comment

viveksinghggits replied:
Location has the information about the object storage that is going to be used by Kanister to upload the backup objects.

Kartik-Garg replied:
Thanks for the comment.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
-	KeyPair           *KeyPair           `json:"keyPair,omitempty"`
-	Secret            *ObjectReference   `json:"secret,omitempty"`
+	// Type of the credential being used.
+	Type CredentialType `json:"type"`
```
Type has the information about how the credentials are going to be provided for the respective object storage.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
+	LocationTypeKopia LocationType = "kopia"
 )
 
 // Location
```
let's remove this comment if we don't need this. if we need this, let's make it better.

Kartik-Garg replied:
I think we might not need these comments, not really sure, as these fields are self explanatory. 
please, do let me know what you think

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
-	Prefix   string       `json:"prefix"`
-	Region   string       `json:"region"`
+	// Type of the Location being used.
+	Type LocationType `json:"type"`
```
Type specifies, the kind of object storage that would be used to upload the backup objects.

viveksinghggits replied:
Let's also specify what are all the types of locations supported. maybe refer to them somehow.

Kartik-Garg replied:
How does this comment sound like :
Type specifies the kind of object storage that would be used to upload the backup objects. For example: "GCS" or "S3Compliant" or "Azure" or "Kopia" .

viveksinghggits replied:
instead of for example I think we should say, currently supported values are.

Kartik-Garg replied:
Alright noted.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
+	// Represents the bucket being used for the current Location.
+	Bucket string `json:"bucket"`
+	// Represents the endpoints being used by the current Location.
+	Endpoint string `json:"endpoint"`
```
Endpoint specifies the endpoint where the object storage is accessible at.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,77 +254,112 @@ type BlueprintPhase struct {
+	// Represents the endpoints being used by the current Location.
+	Endpoint string `json:"endpoint"`
+	// Prefix is the string used in the beginning of the Location.
+	Prefix string `json:"prefix"`
```
I am not sure about this, I will get back to you about this.

Kartik-Garg replied:
Alright, I have added the rest of comments.

viveksinghggits replied:
`Prefix is the string that would be appended into the object path in your bucket where the objects will be uploaded to.`

Kartik-Garg replied:
Thanks for the response.

viveksinghggits replied:
appended or pre pended I am not sure. Can you please check what would be the correct work. The object path would start with this string basically. So this is how the path would look like
```
prefixtest/mysql-backups/mysql/mysql-release/2023-01-03T11-44-24/dump.sql.gz
```

Kartik-Garg replied:
I think the right term is `prepend`.
After changes the comment would look like: 
`Prefix is the string that would be prepended into the object path in the bucket where the objects will be uploaded to.`

viveksinghggits commented:
@Kartik-Garg have you addressed all the review comment's suggestion in this PR? If yes, can you please mark the comments resolved.

Kartik-Garg commented:
> @Kartik-Garg have you addressed all the review comment's suggestion in this PR? If yes, can you please mark the comments resolved.

@viveksinghggits There's just one comment left about Prefix, other than that I have made changes and commited

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -51,8 +51,10 @@ var _ runtime.Object = (*ActionSet)(nil)
 	metav1.ObjectMeta `json:"metadata"`
-	Spec              *ActionSetSpec   `json:"spec,omitempty"`
-	Status            *ActionSetStatus `json:"status,omitempty"`
+	// Spec defines the specification for ActionSet
```
```suggestion
	// Spec defines the specification for ActionSet, for example which action should be performed and with which blueprint etc. It also has details about the profile and any other object that might be used by referred blueprint.
```

viveksinghggits replied:
feel free to make these sentences better with the help of grammar checker.

Kartik-Garg replied:
Alright!

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -109,9 +112,14 @@ type ActionSpec struct {
-	Actions  []ActionStatus `json:"actions,omitempty"`
-	Error    Error          `json:"error,omitempty"`
+	// State represents the current state of the ActionSet.
+	State State `json:"state"`
```
Let's also mention all the possible values for State here. Let me know if you need some help.

Kartik-Garg replied:
Are the state values: Pending, Running, Failed and complete?

viveksinghggits replied:
yes

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -156,21 +164,27 @@ const (
 
+// Error is used to show error messages occured.
 type Error struct {
+	// Message displayed in case if error occurs.
```
```suggestion
	// Message is the actual error message that is displayed in case of errors.
```

Kartik-Garg replied:
Changed accordingly

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +211,42 @@ var _ runtime.Object = (*Blueprint)(nil)
-	OutputArtifacts    map[string]Artifact `json:"outputArtifacts,omitempty"`
-	Phases             []BlueprintPhase    `json:"phases,omitempty"`
-	DeferPhase         *BlueprintPhase     `json:"deferPhase,omitempty"`
+	// Name of the action.
```
```suggestion
	// Name contains the name of the action.
```

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +211,42 @@ var _ runtime.Object = (*Blueprint)(nil)
-	DeferPhase         *BlueprintPhase     `json:"deferPhase,omitempty"`
+	// Name of the action.
+	Name string `json:"name"`
+	// Kind field is now deprecated.
```
```suggestion
	// Kind contains the resource this actions is to be performed on. Kind field is now deprecated.
```

Kartik-Garg replied:
worded it like `Kind contains the resource on which this action has to be performed. Kind field is now deprecated.`

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +211,42 @@ var _ runtime.Object = (*Blueprint)(nil)
-	Name       string                     `json:"name"`
+	// Func is the name of a registered Kanister function.
+	Func string `json:"func"`
+	// Name of the phase.
```
```suggestion
	// Name contains name of the phase.
```

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -109,9 +112,14 @@ type ActionSpec struct {
-	Error    Error          `json:"error,omitempty"`
+	// State represents the current state of the ActionSet.
+	State State `json:"state"`
+	// Actions represents the latest available observations of an ActionSet's current state.
```
Made the changes accordingly

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +211,42 @@ var _ runtime.Object = (*Blueprint)(nil)
+	Func string `json:"func"`
+	// Name of the phase.
+	Name string `json:"name"`
+	// ObjectRefs represnets a map of references to the Kubernetes objects on which the action will be performed.
```
```suggestion
	// ObjectRefs represnets a map of references to the Kubernetes objects that can later be used in the `Args` of the function.
```

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,20 +255,27 @@ type BlueprintPhase struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// Profile
+// Profile captures information about a location for data operation artifacts and
+// corresponding credentials that will be made available to a Blueprint.
```
```suggestion
// Profile captures information about, a location for data operation artifacts and
// corresponding credentials, that will be made available to a Blueprint.
```

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -225,20 +255,27 @@ type BlueprintPhase struct {
+	Credential Credential `json:"credential"`
+	// SkipSSLVerify is boolean and specifies whether skipping SkipSSLVerify verification
+	// is allowed when operating with the Location.
+	// If omitted from a CR definition it default to false
```
```suggestion
	// If omitted from a CR definition it defaults to false
```

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -253,49 +290,73 @@ const (
+	CredentialTypeKopia CredentialType = "kopia"
 )
 
 // Credential
```
If this comment is not necessary, let's remove it.

viveksinghggits replied:
I don't think we can just remove it, go lint my fail. So before committing it please make sure go run `make golint `

Kartik-Garg replied:
I think the fields are self-explanatory. So, removing the comments.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -253,49 +290,73 @@ const (
-	IDField     string          `json:"idField"`
-	SecretField string          `json:"secretField"`
-	Secret      ObjectReference `json:"secret"`
+	// IDField represents the IDs of corresponding keys in the Secret under which KeyPair
```
```suggestion
	// IDField represents the IDs of corresponding keys in the Secret under which KeyPair
```
IDs or ID?

Kartik-Garg replied:
Since this field can be used to store IDs of multiple keys, so maybe IDs is correct?

viveksinghggits commented:
Looks in good shape.
@pavannd1 can you please have a look quick look when you get time. And then we can merge this.

viveksinghggits commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -196,27 +212,42 @@ var _ runtime.Object = (*Blueprint)(nil)
-	DeferPhase         *BlueprintPhase     `json:"deferPhase,omitempty"`
+	// Name contains the name of the action.
+	Name string `json:"name"`
+	// Kind contains the resource on which this action has to be performed. Kind field is now deprecated.
```
Let's remove second sentence from here. We still use this in our blueprints. And if we decide that we want to remove that from blueprint we can update this.

Kartik-Garg replied:
Made the changes.

viveksinghggits commented:
LGTM. I will approve once Pavan has taken a look.

pavannd1 reviewed (commented):
Part 1 of my review. I'll post the remaining comments tomorrow

Kartik-Garg commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -51,8 +51,11 @@ var _ runtime.Object = (*ActionSet)(nil)
 	metav1.ObjectMeta `json:"metadata"`
-	Spec              *ActionSetSpec   `json:"spec,omitempty"`
-	Status            *ActionSetStatus `json:"status,omitempty"`
+	// Spec defines the specification for ActionSet. For example,
+	// which action should be performed with which blueprint, etc. It also has details about the profile and other objects that might be used by the referred blueprint.
```
Just sticking to `actionset` then.

Kartik-Garg commented:
> Part 1 of my review. I'll post the remaining comments tomorrow

Thanks a lot for the reviews, made changes according to the review comments, eagerly waiting for upcoming parts :D

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -198,27 +217,42 @@ var _ runtime.Object = (*Blueprint)(nil)
+	// Phases is the list of BlueprintPhases which are invoked in order when executing this action.
+	Phases []BlueprintPhase `json:"phases,omitempty"`
+	// DeferPhase is invoked after the execution of Phases that are defined for an action.
+	// A DeferPhase, is executed regardless of the statuses of the other phases of the action.
```
```suggestion
	// A DeferPhase is executed regardless of the statuses of the other phases of the action.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -198,27 +217,42 @@ var _ runtime.Object = (*Blueprint)(nil)
+	Func string `json:"func"`
+	// Name contains name of the phase.
+	Name string `json:"name"`
+	// ObjectRefs represnets a map of references to the Kubernetes objects that can later be used in the `Args` of the function.
```
```suggestion
	// ObjectRefs represents a map of references to the Kubernetes objects that
	// can later be used in the `Args` of the function.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -227,20 +261,27 @@ type BlueprintPhase struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// Profile
+// Profile captures information about, a location for data operation artifacts and
+// corresponding credentials, that will be made available to a Blueprint.
```
```suggestion
// Profile captures information about a storage location for backup artifacts and
// corresponding credentials, that will be made available to a Blueprint phase.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -227,20 +261,27 @@ type BlueprintPhase struct {
+	Location Location `json:"location"`
+	// Credential represents the credentials associated with the Location.
+	Credential Credential `json:"credential"`
+	// SkipSSLVerify is boolean and specifies whether skipping SkipSSLVerify verification
+	// is allowed when operating with the Location.
+	// If omitted from a CR definition it defaults to false
```
```suggestion
	// SkipSSLVerify is a boolean that specifies whether skipping SSL verification
	// is allowed when operating with the Location.
	// If omitted from the CR definition, it defaults to false
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -255,11 +296,16 @@ const (
-	Endpoint string       `json:"endpoint"`
-	Prefix   string       `json:"prefix"`
-	Region   string       `json:"region"`
+	// Type specifies the kind of object storage that would be used to upload the backup objects. Currently supported values are: "GCS", "S3Compliant", "Azure" and "Kopia".
```
```suggestion
	// Type specifies the kind of object storage that would be used to upload the
	// backup objects. Currently supported values are: "GCS", "S3Compliant",
	// and "Azure".
```
Let's not add `Kopia`

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -255,11 +296,16 @@ const (
-	Region   string       `json:"region"`
+	// Type specifies the kind of object storage that would be used to upload the backup objects. Currently supported values are: "GCS", "S3Compliant", "Azure" and "Kopia".
+	Type LocationType `json:"type"`
+	// Bucket represents the bucket on the object storage where the backup would be uploaded to.
```
```suggestion
	// Bucket represents the bucket on the object storage where the backup is uploaded.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -255,11 +296,16 @@ const (
+	Bucket string `json:"bucket"`
+	// Endpoint specifies the endpoint where the object storage is accessible at.
+	Endpoint string `json:"endpoint"`
+	// Prefix is the string that would be prepended into the object path in the bucket where the objects will be uploaded to.
```
```suggestion
	// Prefix is the string that would be prepended to the object path in the
	// bucket where the backup objects are uploaded.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -255,11 +296,16 @@ const (
+	Endpoint string `json:"endpoint"`
+	// Prefix is the string that would be prepended into the object path in the bucket where the objects will be uploaded to.
+	Prefix string `json:"prefix"`
+	// Region represents the region where the bucket mentioned above is present in.
```
```suggestion
	// Region represents the region of the bucket specified above.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
-	Type              CredentialType     `json:"type"`
-	KeyPair           *KeyPair           `json:"keyPair,omitempty"`
-	Secret            *ObjectReference   `json:"secret,omitempty"`
+	// Type represents the information about how the credentials are going to be provided for the respective object storage.
```
```suggestion
	// Type represents the information about how the credentials are provided for the respective object storage.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
-	Secret            *ObjectReference   `json:"secret,omitempty"`
+	// Type represents the information about how the credentials are going to be provided for the respective object storage.
+	Type CredentialType `json:"type"`
+	// KeyPair represnets the pair of Keys being used for the Credential of Type KeyPair.
```
```suggestion
	// KeyPair represents the key-value map used for the Credential of Type KeyPair.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
-	IDField     string          `json:"idField"`
-	SecretField string          `json:"secretField"`
-	Secret      ObjectReference `json:"secret"`
+	// IDField represents the IDs of corresponding keys in the Secret under which KeyPair
+	// Credentials are stored.
```
```suggestion
	// IDField specifies the corresponding key in the secret where the AWS Key ID value is stored.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
+	// IDField represents the IDs of corresponding keys in the Secret under which KeyPair
+	// Credentials are stored.
+	IDField string `json:"idField"`
+	// SecretField specifies the corresponding keys in the Secret under which KeyPair
+	// Credentials are stored.
```
```suggestion
	// SecretField specifies the corresponding key in the secret where the AWS Secret Key value is stored.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
 type KopiaServerSecret struct {
-	Username       string                `json:"username,omitempty"`
-	Hostname       string                `json:"hostname,omitempty"`
+	// Username represents the UserName used to connect to the Kopia Server.
```
```suggestion
	// Username represents the username used to connect to the Kopia Server.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
-	Hostname       string                `json:"hostname,omitempty"`
+	// Username represents the UserName used to connect to the Kopia Server.
+	Username string `json:"username,omitempty"`
+	// Hostname rpresents the Hostname used to connect to the Kopia Server.
```
```suggestion
	// Hostname represents the hostname used to connect to the Kopia Server.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
 // KopiaServerSecretRef refers to K8s secrets containing Kopia creds
 type KopiaServerSecretRef struct {
-	Key    string           `json:"key"`
+	// Key is part of K8s secrets that is used to access Kopia Server.
```
```suggestion
	// Key represents the corresponding key in the secret where the required
	// credential or certificate value is stored.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -273,31 +319,47 @@ const (
-	Key    string           `json:"key"`
+	// Key is part of K8s secrets that is used to access Kopia Server.
+	Key string `json:"key"`
+	// Secret is part of K8s secrets that is used to access Kopia Server.
```
```suggestion
	// Secret is the K8s secret object where the creds related to the Kopia Server are stored.
```

pavannd1 commented on pkg/apis/cr/v1alpha1/types.go:
```diff
@@ -307,5 +369,6 @@ type KopiaServerSecretRef struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata"`
-	Items           []*Profile `json:"items"`
+	// Items represnets a list of Profiles.
```
```suggestion
	// Items represents a list of Profiles.
```

Kartik-Garg commented:
@pavannd1 Thanks for the review comments, I have made the commits according to the suggestion. :D

pavannd1 reviewed (approved):
LGTM

Kartik-Garg commented:
@pavannd1 I just updated the branch with respect to upstream, if everything is in order, could you please merge the PR
